### PR TITLE
Fix `segmented_gather()` for null LIST rows

### PR DESCRIPTION
**Pull request by mythrocks**
`segmented_gather()` currently assumes that null LIST rows also have
a `0` size (as defined by the difference of adjacent offsets.)

This might not hold, for example, for LIST columns that are members
of STRUCT columns whose parent null masks are superimposed on its
children. This would cause a non-empty list row to be marked null,
without compaction. This leads to errors in fetching elements of a
list row as seen in NVIDIA/spark-rapids/pull/3770.

This commit adds the handling of uncompacted LIST rows in
`segmented_gather()`.

### CODE REVIEW

File: cpp/tests/copying/segmented_gather_list_tests.cpp [hunk 1]
```diff
@@ -300,6 +299,29 @@ TYPED_TEST(SegmentedGatherTest, GatherNegatives)
...
+  cudf::detail::set_null_mask(input->mutable_view().null_mask(), 5, 6, false);
+
+  auto const gather_map = LCW<int>{{-1, 2, 1, -4}, {0}, {-2, 1}, {0, 2, 1}, {}, {0}, {1, 2}};
+  auto const expected =
+    LCW<T>{{{4, 3, 2, 1}, {5}, {6, 7}, {8, 0, 9}, {}, {{X}, all_nulls()}, {4, 5}}, null_at(5)};
+  auto const results = segmented_gather(lists_column_view{*input}, lists_column_view{gather_map});
+  std::cout << "Results:  " << std::endl;
+  cudf::test::print(*results);
+  std::cout << "Expected: " << std::endl;
+  cudf::test::print(expected);
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply cpp/tests/copying/segmented_gather_list_tests.cpp:321]: Argh! Thanks for catching that.

**Comment by davidwendt** [comment]
Could you run this with `compute-sanitizer`?
```
compute-sanitizer --tool memcheck gtests/COPYING_TEST --gtest_filter=SegmentedGatherTest* --rmm_mode=cuda
```

Same for the extract tests too.

File: cpp/tests/copying/segmented_gather_list_tests.cpp [hunk 2]
```diff
@@ -44,9 +45,7 @@ template <typename T>
 using LCW = cudf::test::lists_column_wrapper<T, int32_t>;
 using cudf::lists_column_view;
 using cudf::lists::detail::segmented_gather;
-using cudf::test::iterators::no_nulls;
-using cudf::test::iterators::null_at;
-using cudf::test::iterators::nulls_at;
+using namespace cudf::test::iterators;
```
Comment (davidwendt) [inline cpp/tests/copying/segmented_gather_list_tests.cpp:48]: Could you remove the `class SegmentedGatherTestList` from above? It does not appear to be used anywhere.

**Comment by mythrocks** [comment]
> Could you run this with `compute-sanitizer`?
> 
> ```
> compute-sanitizer --tool memcheck gtests/COPYING_TEST --gtest_filter=SegmentedGatherTest* --rmm_mode=cuda
> ```
> 
> Same for the extract tests too.

Thank you for suggesting the `memcheck` test:
```
$ git:(extract-list-element-nulls) ✗ compute-sanitizer --tool memcheck gtests/COPYING_TEST --gtest_filter=SegmentedGatherTest\* --rmm_mode=cuda
...

[----------] Global test environment tear-down
[==========] 203 tests from 22 test suites ran. (142742 ms total)
[  PASSED  ] 203 tests.
========= ERROR SUMMARY: 0 errors

$ git:(extract-list-element-nulls) ✗ compute-sanitizer --tool memcheck gtests/LISTS_TEST --gtest_filter=ListsExtract\* --rmm_mode=cuda 
...
[----------] Global test environment tear-down
[==========] 80 tests from 35 test suites ran. (25638 ms total)
[  PASSED  ] 80 tests.
========= ERROR SUMMARY: 0 errors
```

File: cpp/tests/copying/segmented_gather_list_tests.cpp (hunk 2 above)
Comment (mythrocks) [reply cpp/tests/copying/segmented_gather_list_tests.cpp:45]: Done.

**Review by ttnghia — APPROVED** [review]

**Review by davidwendt — APPROVED** [review]

**Comment by davidwendt** [comment]
rerun tests

**Comment by mythrocks** [comment]
Rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by mythrocks** [comment]
@gpucibot merge

**Comment by mythrocks** [comment]
@davidwendt, @codereport, @ttnghia: Thank you for the reviews.
@galipremsagar: Thank you for sorting out the 11.5 CI failure.

I have merged this change.

**Pull request merged**